pkg/util/kubernetes/apiserver/controllers/metadata_controller_test.go

```diff
@@ -310,6 +363,304 @@ func TestMetadataControllerSyncEndpoints(t *testing.T) {
+						Name:            "svc2-slice1",
+						Labels:          map[string]string{"kubernetes.io/service-name": "svc2"},
+						ResourceVersion: "v1",
+					},
+					Endpoints: []discv1.Endpoint{
+						newFakeEndpoint("node1", pod1),
+					},
+				},
+			},
+			map[string]apiv1.NamespacesPodsStringsSet{
+				"node1": {
+					"default": {
+						"pod1_name": sets.New("svc1", "svc2"),
+					},
+				},
+				"node2": {
+					"default": {
+						"pod2_name": sets.New("svc1"),
+					},
+				},
+				"node3": {
+					"default": {
+						"pod3_name": sets.New("svc1"),
+					},
+				},
+			},
+			3, // Cache length increases by 1 for svc2-slice1
+		},
+		{
+			"update slice to add pod",
+			false,
+			[]*discv1.EndpointSlice{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace:       "default",
+						Name:            "svc2-slice1",
+						Labels:          map[string]string{"kubernetes.io/service-name": "svc2"},
+						ResourceVersion: "v2",
+					},
+					Endpoints: []discv1.Endpoint{
+						newFakeEndpoint("node1", pod1),
+						newFakeEndpoint("node2", pod2),
+					},
+				},
+			},
+			map[string]apiv1.NamespacesPodsStringsSet{
+				"node1": {
+					"default": {
+						"pod1_name": sets.New("svc1", "svc2"),
+					},
+				},
+				"node2": {
+					"default": {
+						"pod2_name": sets.New("svc1", "svc2"),
+					},
+				},
+				"node3": {
+					"default": {
+						"pod3_name": sets.New("svc1"),
+					},
+				},
+			},
+			3,
+		},
+		{
+			"update slice to delete pod",
+			false,
+			[]*discv1.EndpointSlice{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace:       "default",
+						Name:            "svc2-slice1",
+						Labels:          map[string]string{"kubernetes.io/service-name": "svc2"},
+						ResourceVersion: "v3",
+					},
+					Endpoints: []discv1.Endpoint{
+						newFakeEndpoint("node2", pod2),
+					},
+				},
+			},
+			map[string]apiv1.NamespacesPodsStringsSet{
+				"node1": {
+					"default": {
+						"pod1_name": sets.New("svc1"),
+					},
+				},
+				"node2": {
+					"default": {
+						"pod2_name": sets.New("svc1", "svc2"),
+					},
+				},
+				"node3": {
+					"default": {
+						"pod3_name": sets.New("svc1"),
+					},
+				},
+			},
+			3,
+		},
+		{
+			"delete service with multiple slices",
+			true,
+			[]*discv1.EndpointSlice{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace:       "default",
+						Name:            "svc1-slice1",
+						Labels:          map[string]string{"kubernetes.io/service-name": "svc1"},
+						ResourceVersion: "v2",
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace:       "default",
+						Name:            "svc1-slice2",
+						Labels:          map[string]string{"kubernetes.io/service-name": "svc1"},
+						ResourceVersion: "v2",
+					},
+				},
+			},
+			map[string]apiv1.NamespacesPodsStringsSet{
+				"node2": {
+					"default": {
+						"pod2_name": sets.New("svc2"),
+					},
+				},
+			},
+			1,
+		},
+	}
+
+	for i, tt := range tests {
```

Comment:
nit: can we run the tests using the following pattern. will be more clear output for debugging when subtests fail.
```
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
                        ... <test_contents> ...
		})
	}
```